### ISSUE TITLE
tr will ignore "string" prefixes if it finds a match inside a str().

### ISSUE THREAD

**Issue by TheFoxKnocks** [godotengine/godot#80501]
### Godot version

v4.1.stable.official [970459615]

### System information

Godot v4.1.stable - Windows 10.0.19045 - Vulkan (Compatibility) - NVIDIA GeForce RTX 3080 (NVIDIA; 31.0.15.3179) - 12th Gen Intel(R) Core(TM) i9-12900KF (24 Threads)

### Issue description

I'm using Resources, and inside these Resources are 3 tags. Within my localization .CSV, all items which could be used as a tag are prefixed with "tag_", so, for example... "tag_area" is a key which translates into "AREA".

So, if I put the tag holding "area" into a variable, such as "tag3":

tr("tag_" + str(tag3))

What will happen is that it will decide to not grab "tag_area", but instead just grab "area".

All of my other tags work fine with this system, stuff like "tag_fire" or "tag_directional", but that's because there isn't a non-tag version. There is only "tag_fire" and "fire" does not exist. But I DO have "tag_area" and "area". It's pulling the wrong one.

### Steps to reproduce

Create resource that has a variable holding a string.

Create a .CSV Localization file with a key that includes text from within that string.

Print it in code doing something like tr("key_first_half_" + str(key_second_half))" and it will print the result of key_second_half if that also exists as a key, ignoring the key_first_half_ prefix.

i.e; if you have 2 keys, and one is called "key_first_half_test" and the key_second_half variable is holding a string called "test", if your other key is just called "test" it'll mistakenly grab that and pretend the "key_first_half_" part doesn't exist.

### Minimal reproduction project

N/A

**Comment by Calinou** [comment]
@TheFoxKnocks Please upload a [minimal reproduction project](https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#provide-a-simple-example-project) to make this easier to troubleshoot.

PS: It's usually not recommended to programmatically construct translation keys, as this makes it impossible for translation extraction tools to write out files with all translations keys.

**Comment by TheFoxKnocks** [comment]
> 

What is the recommended way, then?

**Comment by dalexeev** [comment]
I don't think this is possible for variable arguments. Support for constant arguments has been added recently though, see #80020.

If you want to substitute some value in a string, you can use the `%` operator or the `format()` method (the latter is preferable, since `format()` supports named keys and allows a set of substitution keys and values to not match).

```gdscript
tr("Hello, %s!") % player_name
tr("{player} get {item}.").format({player = player_name, item = item_name})
```

If you use some technical IDs that you generate from variable expressions as translation keys (like `tr("ITEM_" + item_id)`), then you can either collect a list of possible keys manually or make a plugin (`EditorTranslationParserPlugin` for extracting strings and/or `ResourceFormatLoader`/`EditorImportPlugin` for `Translation` resource with `_get_message()` overriden).

**Comment by timothyqiu** [comment]
Ah, after being confused for two years, I think I finally understand this issue.

This is the key:

> All of my other tags work fine with this system, stuff like "tag_fire" or "tag_directional", but that's because there isn't a non-tag version. There is only "tag_fire" and "fire" does not exist. But I DO have "tag_area" and "area". It's pulling the wrong one.

1. OP used `"tag_" + str(tag3)` to make the key `tag_area`.
2. OP used `tr()` to turn `tag_area` into `area`.
3. As auto translation is enabled by default, and there happens to be a translation for `area`, the displayed text is turned into `Translation for area`.

So the solution to the problem is to either:

- Use `label.text = "tag_" + str(tag3)` instead of `label.text = tr("tag_" + str(tag3))`, or
- Turn off auto translation for the label as you are doing translation manually via `tr()`

Closing since this thread has been quiet for quite some time. Feel free to reopen or submit a new one if you have further questions.
